numba_dpex/core/caching.py

```diff
@@ -0,0 +1,387 @@
+                The target context for the kernel.
+            reduced_data (object): The data to be deserialzed after unpickling.
+        """
+        # TODO: Implement, but looks like we might not need it at all.
+        # Look at numba.core.caching for how to implement.
+        pass
+
+    def check_cachable(self, cres):
+        """Check if a certain object is cacheable.
+
+        Args:
+            cres (object): The object to be cached. For example, if the object is
+            `CompileResult`, then you might want to follow the similar checks as
+            has been done in `numba.core.caching.CompileResultCacheImpl`.
+
+        Returns:
+            bool: Return `True` if cacheable, otherwise `False`.
+        """
+        # TODO: Although, for the time being, assuming all Kernels in numba_dpex
+        # are always cachable. However, we might need to add some bells and
+        # whistles in the future. Look at numba.core.caching for how to implement.
+        return True
+
+
+class AbstractCache(metaclass=ABCMeta):
+    @abstractmethod
+    def get(self):
+        """_summary_"""
+
+    @abstractmethod
+    def put(self, key, value):
+        """_summary_
+
+        Args:
+            key (_type_): _description_
+            value (_type_): _description_
+
+        Returns:
+            _type_: _description_
+        """
+
+
+class NullCache(AbstractCache):
+    def get(self, key):
+        return None
+
+    def put(self, key, value):
+        pass
+
+
+class Node:
+    """A 'Node' class for LRUCache."""
+
+    def __init__(self, key, value):
+        """Constructor for the Node
+
+        Args:
+            key (object): The key to the value.
+            value (object): The data to be saved.
+        """
+        self.key = key
+        self.value = value
+        self.next = None
+        self.previous = None
+
+    def __str__(self):
+        """__str__ for Node
+
+        Returns:
+            str: str: A human readable representation of a Node.
+        """
+        return "(" + str(self.key) + ": " + str(self.value) + ")"
+
+    def __repr__(self):
+        """__repr__ for Node
+
+        Returns:
+            sstr: A human readable representation of a Node.
+        """
+        return self.__str__()
+
+
+class LRUCache(AbstractCache):
+    """LRUCache implementation for caching kernels, functions and modules
+
+    The cache is basically a doubly-linked-list backed with a dictionary
+    as a lookup table.
+    """
+
+    def __init__(self, capacity=10, pyfunc=None):
+        """Constructor for LRUCache
+
+        Args:
+            capacity (int, optional): The max capacity of the cache. Defaults to 10.
+            pyfunc (NoneType, optional): A python function to be cached. Defaults to None.
+        """
+        self._capacity = capacity
+        self._lookup = {}
+        self._evicted = {}
+        self._dummy = Node(0, 0)
+        self._head = self._dummy.next
```

Comment:
Where are we setting `_dummy.next`? The `Node` class does not have a `next` attribute.

Reply:
The `Node` class has `next` and `previous`.